homeassistant/components/switch/mochad.py

```diff
@@ -59,17 +64,37 @@ def name(self):
 
     def turn_on(self, **kwargs):
         """Turn the switch on."""
-        self._state = True
+        from pymochad.exceptions import MochadException
+        _LOGGER.debug("Reconnect %s:%s", self._controller.server,
+                      self._controller.port)
         with mochad.REQ_LOCK:
-            self.device.send_cmd('on')
-            self._controller.read_data()
+            try:
+                # Recycle socket on new command to recover mochad connection
+                self._controller.reconnect()
+                self.device.send_cmd('on')
+                # No read data on CM19A which is rf only
+                if self._comm_type == 'pl':
+                    self._controller.read_data()
```

Comment:
Why is this done without handling the return value? Shouldn't the return value be used to update state?

Reply:
Original code has read_data() after each send_cmd(). This works for pl, but not for rf. 
pl = powerline devices (like appliance switches). rf = wireless devices (mostly sensors).  When sending data to pl devices = they respond back with the state. rf devices do not do that. CM19A unlike CM15A, handles all commands as rf. It relies on a separate translator device - TM751 for rf->pl translation. And it's only one way.  There is no read back.  The original code gets in stuck on read_data() for me until any data shows up from sensors in the buffer. Making it conditional fixes the problem without affecting CM15A behaviours.

Reply:
I understand that RF devices don't support this call, but I don't understand why the call is done at all, when the return value is not handled.

Reply:
This call is done for pl devices in order to clear the receive buffer. When pl device is set to On/Off state, it responds back with it's new value.  This read command reads it and ignores it. The state is already known on hass side.

Reply:
Ok, the empty read buffer makes sense, but since the device responds with the new state, that should be used in home assistant. Home assistant doesn't know the new state for sure. Sending a command to the device over some kind of network transport doesn't guarantee that the device receives the command or changes state. We can only be sure after the device has replied to home assistant with its new state. So whenever possible, a device should feedback its state to home assistant when that changes.